File: ICPC/2015 Divisionals/clever/submissions/clever-darcy.cpp
#include <bits/stdc++.h>
using namespace std;

int count(const vector<string>& names,const string& word){
  if(word.length() != names.size()) return false;
  int n = word.length();
  
  vector<int> DP(1 << n);
  DP[0] = 1;
  
  for(int i=1;i<(1 << n);i++){
    char lett = word[__builtin_popcount(i)-1];
    for(int j=0;j<n;j++)
      if(((i >> j) & 1) && names[j].find(lett) != string::npos)
        DP[i] += DP[i ^ (1 << j)];
  }

  return DP[(1 << n)-1];
}

int main(){
  int n,k;
  cin >> n >> k;
  
  vector<string> names(n);
  for(int i=0;i<n;i++)
    cin >> names[i];

  for(int i=0;i<k;i++){
    string word;
    cin >> word;
    cout << count(names,word) << endl;
  }
}

File: ICPC/2015 Divisionals/clever/submissions/CleverDarcy.java
import java.util.Scanner;

public class CleverDarcy {
    public static void main(String[] args){
	new CleverDarcy().run();
    }
    
    void run(){
	Scanner in = new Scanner(System.in);
	int n = in.nextInt();
	int k = in.nextInt();

	String[] names = new String[n];
	for(int i=0;i<n;i++)
	    names[i] = in.next();

	for(int i=0;i<k;i++)
	    System.out.println(count(names,in.next()));
    }

    int count(String[] names,String word){
	if(word.length() != names.length) return 0;
	int n = word.length();
  
	int[] DP = new int[1 << n];
	DP[0] = 1;
	
	for(int i=1;i<(1 << n);i++){
	    char lett = word.charAt(Integer.bitCount(i)-1);
	    for(int j=0;j<n;j++)
		if(((i >> j) & 1) != 0 && names[j].indexOf(lett) != -1)
		    DP[i] += DP[i ^ (1 << j)];
	}
	
	return DP[(1 << n)-1];
    }
}
